### PR TITLE
txn: handle errors returned from Flush

### LINKED ISSUE #52889: Errors returned from Flush are not equally treated as those returned from Commit
## Bug Report

ref #50215

### 1. Minimal reproduce step (Required)

See the case `TestPipelinedDMLDisableRetry`.

```go
func TestPipelinedDMLDisableRetry(t *testing.T) {
	// the case tests that
	// 1. auto-retry for pipelined dml is disabled
	// 2. the write conflict error message returned from a Flush (instead of from Commit) is correct
	require.Nil(t, failpoint.Enable("tikvclient/pipelinedMemDBMinFlushKeys", `return(1)`))
	require.Nil(t, failpoint.Enable("tikvclient/pipelinedMemDBMinFlushSize", `return(1)`))
	require.Nil(t, failpoint.Enable("tikvclient/pipelinedMemDBForceFlushSizeThreshold", `return(1)`))
	defer func() {
		require.Nil(t, failpoint.Disable("tikvclient/pipelinedMemDBMinFlushKeys"))
		require.Nil(t, failpoint.Disable("tikvclient/pipelinedMemDBMinFlushSize"))
		require.Nil(t, failpoint.Disable("tikvclient/pipelinedMemDBForceFlushSizeThreshold"))
	}()
	store := realtikvtest.CreateMockStoreAndSetup(t)
	tk1 := testkit.NewTestKit(t, store)
	tk1.Session().GetSessionVars().InitChunkSize = 1
	tk1.Session().GetSessionVars().MaxChunkSize = 1
	tk2 := testkit.NewTestKit(t, store)
	tk1.MustExec("use test")
	tk2.MustExec("use test")
	tk1.MustExec("drop table if exists t")
	tk1.MustExec("create table t(a int primary key, b int)")
	// we need to avoid inserting *literals* into t, so let t2 be the source table.
	tk1.MustExec("create table t2(a int, b int)")
	tk1.MustExec("insert into t2 values (1, 1), (2, 1)")
	require.Nil(t, failpoint.Enable("tikvclient/beforePipelinedFlush", `pause`))
	tk1.MustExec("set session tidb_dml_type = bulk")
	errCh := make(chan error)
	go func() {
		// we expect that this stmt triggers 2 flushes, each containing only 1 row.
		errCh <- tk1.ExecToErr("insert into t select * from t2 order by a")
	}()
	time.Sleep(500 * time.Millisecond)
	tk2.MustExec("insert into t values (1,2)")
	require.Nil(t, failpoint.Disable("tikvclient/beforePipelinedFlush"))
	err := <-errCh
	require.Error(t, err)
	require.True(t, kv.ErrWriteConflict.Equal(err), fmt.Sprintf("error: %s", err))
	require.ErrorContains(t, err, "tableName=test.t")
}
```

### 2. What did you expect to see? (Required)

Write conflict and assertion errors should be handled.

### 3. What did you see instead (Required)

They are directly returned to user.

### 4. What is your TiDB version? (Required)

Master(6ed307ec244cb8823f3e1df9c76b862db901c216)

### PR DESCRIPTION
**Pull request by ekexium**
### What problem does this PR solve?

Issue Number: close #52889

Problem Summary:

### Check List

Tests 

- [x] Unit test
- [ ] Integration test
- [ ] Manual test (add detailed scripts or steps below)
- [ ] No need to test
  > - [ ] I checked and no code files have been changed.
  > 

Side effects

- [ ] Performance regression: Consumes more CPU
- [ ] Performance regression: Consumes more Memory
- [ ] Breaking backward compatibility

Documentation

- [ ] Affects user behaviors
- [ ] Contains syntax changes
- [ ] Contains variable changes
- [ ] Contains experimental features
- [ ] Changes MySQL compatibility

### Release note

Please refer to [Release Notes Language Style Guide](https://pingcap.github.io/tidb-dev-guide/contribute-to-tidb/release-notes-style-guide.html) to write a quality release note.

```release-note
None
```

### CODE REVIEW

File: pkg/session/session.go [hunk 1]
```diff
@@ -2274,6 +2278,17 @@ func runStmt(ctx context.Context, se *session, s sqlexec.Statement) (rs sqlexec.
 	}
 
 	rs, err = s.Exec(ctx)
+
+	// Pipelined-DML can return assertion error and write conflict here, handle them like we handle
+	// errors returned from commit.
+	if err != nil {
```
Comment (cfzjywxk) [inline pkg/session/session.go:2284]: Should there be a condition check like
```
if s.isPipelinedDML {
    if err != nil {
        err = se.handleAssertionFailure(ctx, err)
    }
} else {
    unrechable or errors.
}
```

Comment (ekexium) [reply pkg/session/session.go:2284]: Could be. It also makes the intention clearer.

File: pkg/session/session.go [hunk 2]
```diff
@@ -2274,6 +2276,20 @@ func runStmt(ctx context.Context, se *session, s sqlexec.Statement) (rs sqlexec.
 	}
 
 	rs, err = s.Exec(ctx)
+
+	// check if is pipelined dml
+	if se.txn.Valid() && se.txn.IsPipelined() {
+		// Pipelined-DML can return assertion error and write conflict here, handle them like we handle
+		// errors returned from commit.
```
Comment (crazycs520) [inline pkg/session/session.go:2283]: Could you add comment to explain why need check here? 

The comment only says `can`, but doesn't explain `why`.

File: pkg/session/session.go [hunk 3]
```diff
@@ -846,7 +846,9 @@ func (s *session) tryReplaceWriteConflictError(oldErr error) (newErr error) {
 	}
 	originErr := errors.Cause(oldErr)
 	inErr, _ := originErr.(*errors.Error)
-	args := inErr.Args()
+	oldArgs := inErr.Args()
+	args := make([]any, len(oldArgs))
+	copy(args, oldArgs)
```
Comment (crazycs520) [inline pkg/session/session.go:851]: Also, could you add some comments about why need a copy here?

**Review by crazycs520 — APPROVED** [review]

**Review by cfzjywxk — APPROVED** [review]

**Pull request merged**